File: platform_tools/android/apps/AndroidKit/src/main/java/org/skia/androidkit/Image.java
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

package org.skia.androidkit;

import android.support.annotation.Nullable;
import java.io.InputStream;
import org.skia.androidkit.Matrix;
import org.skia.androidkit.SamplingOptions;
import org.skia.androidkit.Shader;
import org.skia.androidkit.TileMode;

public class Image {
    private long mNativeInstance;

    /**
     * Construct an Image from encoded (PNG, GIF, etc) data.
     *
     * Returns null for unsupported formats or invalid data.
     */
    public static Image fromEncoded(byte[] encodedData) {
        long nativeImage = nCreate(encodedData);
        return nativeImage != 0
            ? new Image(nativeImage)
            : null;
    }

    /**
     * Construct an Image from an encoded data stream.
     *
     * Returns null for unsupported formats or invalid stream.
     */
    public static Image fromStream(InputStream encodedStream) throws java.io.IOException {
        byte[] encodedData = new byte[encodedStream.available()];
        encodedStream.read(encodedData);

        return fromEncoded(encodedData);
    }

    public int getWidth() {
        return nGetWidth(mNativeInstance);
    }

    public int getHeight() {
        return nGetHeight(mNativeInstance);
    }

    public Shader makeShader(TileMode tmx, TileMode tmy, SamplingOptions sampling) {
        return makeShader(tmx, tmy, sampling, null);
    }

    public Shader makeShader(TileMode tmx, TileMode tmy, SamplingOptions sampling,
                             @Nullable Matrix localMatrix) {
        long nativeMatrix = localMatrix != null ? localMatrix.getNativeInstance() : 0;
        return new Shader(nMakeShader(mNativeInstance, tmx.nativeInt, tmy.nativeInt,
                                      sampling.getNativeDesc(),
                                      sampling.getCubicCoeffB(), sampling.getCubicCoeffC(),
                                      nativeMatrix));
    }

    /**
     * Releases any resources associated with this Paint.
     */
    public void release() {
        nRelease(mNativeInstance);
        mNativeInstance = 0;
    }

    @Override
    protected void finalize() throws Throwable {
        release();
    }

    // package private
    Image(long nativeInstance) {
        mNativeInstance = nativeInstance;
    }

    // package private
    long getNativeInstance() { return mNativeInstance; }

    private static native long nCreate(byte[] encodedData);
    private static native void nRelease(long nativeInstance);

    private static native int nGetWidth(long nativeInstance);
    private static native int nGetHeight(long nativeInstance);

    private static native long nMakeShader(long nativeInstance, int tmx, int tmy, int samplingDesc,
                                           float samplingCoeffB, float samplingCoeffC,
                                           long nativeMatrix);
}

File: modules/androidkit/src/Image.cpp
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include <jni.h>

#include "include/core/SkImage.h"
#include "include/core/SkTileMode.h"
#include "modules/androidkit/src/Utils.h"

namespace {

jlong Image_Create(JNIEnv* env, jobject, jbyteArray jdata) {
    auto  size = env->GetArrayLength(jdata);
    auto* data = env->GetByteArrayElements(jdata, nullptr);
    auto image = SkImage::MakeFromEncoded(SkData::MakeWithCopy(data, SkToSizeT(size)));

    env->ReleaseByteArrayElements(jdata, data, 0);

    return reinterpret_cast<jlong>(image.release());
}

void Image_Release(JNIEnv*, jobject, jlong native_instance) {
    SkSafeUnref(reinterpret_cast<const SkImage*>(native_instance));
}

jint Image_GetWidth(JNIEnv*, jobject, jlong native_instance) {
    const auto* image = reinterpret_cast<const SkImage*>(native_instance);
    return image ? image->width() : 0;
}

jint Image_GetHeight(JNIEnv*, jobject, jlong native_instance) {
    const auto* image = reinterpret_cast<const SkImage*>(native_instance);
    return image ? image->height() : 0;
}

jlong Image_MakeShader(JNIEnv*, jobject, jlong native_instance, jint jtmx, jint jtmy,
                       jint sampling_desc, jfloat sampling_b, jfloat sampling_c,
                       jlong native_matrix) {
    sk_sp<SkShader> shader;

    if (const auto* image = reinterpret_cast<const SkImage*>(native_instance)) {
        const auto tmx = androidkit::utils::TileMode(jtmx),
                   tmy = androidkit::utils::TileMode(jtmy);
        const auto sampling = androidkit::utils::SamplingOptions(sampling_desc,
                                                                 sampling_b, sampling_c);

        const auto* lm = reinterpret_cast<const SkM44*>(native_matrix);
        shader = lm
            ? image->makeShader(tmx, tmy, sampling, lm->asM33())
            : image->makeShader(tmx, tmy, sampling);
    }

    return reinterpret_cast<jlong>(shader.release());
}

} // namespace

int register_androidkit_Image(JNIEnv* env) {
    static const JNINativeMethod methods[] = {
        {"nCreate"    , "([B)J"      , reinterpret_cast<void*>(Image_Create)   },
        {"nRelease"   , "(J)V"       , reinterpret_cast<void*>(Image_Release)  },

        {"nGetWidth"  , "(J)I"       , reinterpret_cast<void*>(Image_GetWidth) },
        {"nGetHeight" , "(J)I"       , reinterpret_cast<void*>(Image_GetHeight)},

        {"nMakeShader", "(JIIIFFJ)J" , reinterpret_cast<void*>(Image_MakeShader)},
    };

    const auto clazz = env->FindClass("org/skia/androidkit/Image");
    return clazz
        ? env->RegisterNatives(clazz, methods, SK_ARRAY_COUNT(methods))
        : JNI_ERR;
}
